### PR TITLE
proxy: Improve access log error handling and debug messages

### PR DESCRIPTION
**Pull request by tgraf**
Signed-off-by: Thomas Graf <thomas@cilium.io>

### CODE REVIEW

File: pkg/proxy/log.go [hunk 1]
```diff
@@ -35,12 +51,21 @@ func OpenLogfile(lf string) error {
 		return err
 	}
 
+	logPath = lf
+	log.WithFields(log.Fields{
+		fieldPath: logPath,
+	}).Debugf("Opened access log")
```
Comment (rlenglet) [inline pkg/proxy/log.go]: Debugf -> Debug

File: pkg/proxy/log.go [hunk 2]
```diff
@@ -35,12 +51,21 @@ func OpenLogfile(lf string) error {
...
+
 	logBuf = bufio.NewWriter(logFile)
 	return nil
 }
 
 // CloseLogfile closes the logfile
 func CloseLogfile() {
+	log.WithFields(log.Fields{
+		fieldPath: logPath,
+	}).Debugf("Closed access log")
```
Comment (rlenglet) [inline pkg/proxy/log.go]: Debugf -> Debug

File: pkg/proxy/log.go [hunk 3]
```diff
@@ -68,12 +111,34 @@ func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
...
+	log.WithFields(log.Fields{
+		fieldType:     typ,
+		fieldVerdict:  verdict,
+		fieldCode:     code,
+		fieldMethod:   l.request.Method,
+		fieldURL:      l.request.URL,
+		fieldProtocol: l.request.Proto,
+		fieldHeader:   l.request.Header,
+		fieldPath:     logPath,
+	}).Debugf("logging L7 flow record")
```
Comment (rlenglet) [inline pkg/proxy/log.go]: Debugf -> Debug
logging -> Logging (to be consistent with other log messages)

File: pkg/proxy/proxy.go [hunk 4]
```diff
@@ -380,7 +380,7 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, source Pr
 	gcOnce.Do(func() {
 		if lf := viper.GetString("access-log"); lf != "" {
 			if err := OpenLogfile(lf); err != nil {
-				log.Warningf("cannot open access log: %s", err)
+				log.WithError(err).Warning("cannot open L7 access log")
```
Comment (rlenglet) [inline pkg/proxy/proxy.go]: cannot -> Cannot

Add a field with the log filename?

File: pkg/proxy/log.go [hunk 5]
```diff
@@ -50,11 +75,29 @@ func SetMetadata(md []string) {
...
+
 // Log logs a record to the logfile and flushes the buffer
 func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
-	if logBuf == nil {
-		return
-	}
+	// Lock while writing access log so we serialize writes as we may have
+	// to reopen the logfile and parallel writes could fail because of that
+	logMutex.Lock()
+	defer logMutex.Unlock()
```
Comment (aanm) [inline pkg/proxy/log.go:100]: Can't we use logrus to write directly to the file instead?

something like this (untested code):
```
import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"

	"github.com/Sirupsen/logrus"
)

var (
	log      = logrus.New()
	logBuf   *bufio.Writer
	metadata []string
)

// OpenLogfile opens a file for logging
func OpenLogfile(lf string) error {
	logFile, err := os.OpenFile(lf, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
	if err != nil {
		return err
	}
	log.Out = logFile
	return nil
}
```

Comment (tgraf) [reply pkg/proxy/log.go:100]: What's the advantage of that? Logrus does not even attempt to reopen the logfile.

Comment (aanm) [reply pkg/proxy/log.go:100]: I see, I thought it would be simpler to leave the file writes to logrus

**Review by rlenglet — APPROVED** [review]

**Pull request merged**